src/drivers/cmakeDriver.ts

```diff
@@ -1730,6 +1747,67 @@ export abstract class CMakeDriver implements vscode.Disposable {
         }
     }
 
+    private async _showTaskQuickPick(tasks: vscode.Task[]): Promise<vscode.Task | undefined> {
+        interface TaskItem extends vscode.QuickPickItem {
+            task: vscode.Task;
+        }
+        const choices = tasks.map((t): TaskItem => ({ label: t.name, task: t }));
+        const sel = await vscode.window.showQuickPick(choices, { placeHolder: localize('select.build.task', 'Select build task') });
+        return sel ? sel.task : undefined;
+    }
+
+    private async _findBuildTask(targets?: string[]): Promise<vscode.Task | undefined> {
+
+        // Fetch all CMake task
+        const tasks = (await vscode.tasks.fetchTasks({ type: 'cmake' })).filter(task => ((task.group?.id === vscode.TaskGroup.Build.id)));
+
+        // There is only one found - stop searching
+        if (tasks.length === 1) {
+            return tasks[0];
+        }
+
+        // Get only tasks which target matches to one which is requested
+        const targetTasks = tasks.filter(task =>
+            (task.definition as CMakeTaskDefinition).targets === targets);
```

Comment:
I don't think we should try to match targets.  It would be better if we just replaced the targets with the one selected from the outline or command.  Consider the case where targets === ['all'].  The developer may be surprised that all the targets are being built.

Reply:
That was exactly was I did before (took default task and replace target). I was asked to change it.

I think that current proposal is sensible.
In first place I try finding tasks which targets are matching exactly to requested target. Because that means that there is specific task for this target.
When such task is not found than I try finding tasks without any target.
SO there will be no surprise because case when target X is requested and 'all' target is built should never happen.

Reply:
@bobbrow I think I know what you mean.

I have just discovered that filtering wasn't implemented correctly.
I've fixed filtering but that revealed another thing: all tasks added by CMakeTaskProvider have targets set to ["all"] by default.

So question is what templetae tasks are really expected to do - I suppose that originally, they were planned as template for custom tasks. But at the end they are currently used as "build" tasks if tasks.json have no other cmake tasks.

Reply:
Alternatively CMakeTaskProvider May Just leave targets property as undefined instead of pushing ['all'] into it.
I dont think it is necessary to set it as later on when task is executed it can pull target which is set as default in CMakeProject.
But maybe there are some other reason to ha e it set to ['all']?

Reply:
The reason was to demonstrate that the property exists. We aren't thinking about changing that behavior right now and many people have already generated tasks with this field set at this point.  So we'll need to handle that case.  My vote is to allow "all" to match any target.

This filter statement would need to change anyway because it's expecting the whole targets array to match and what we actually want to do is see if the current target (or "all") is in the array.

Reply:
The algorithm sounds ok.  We don't have statistics on the typical configuration.  Target names are personal/private information and we don't track that.  We could log the frequency which the targets are set and how often "all" is encountered since that's not personally identifiable information, but we hadn't done that yet since we didn't see a need for the information at the time.  That said, most people use defaults and don't change them.

Reply:
Uh... That won't be so easy.
I have just tried implementing mentioned algo.
The problem which arises is that tasks injected by CMakeTaskProvider do not neceserily have "targets" set to ['all'].
They have targets set to ['all'] only if default target (e.g. from status bar) is not set. If it is set then TaskProvider injects default target.
That brings some challenges:
 e.g., imagine tasks.json with single task inside without any targets (let's call it "universal" task). This definition lends as Prio "3".
But tasks added by CMakeTaskProvider have targets defined to "DefaultTarget". When from ProjectOutline I select target which matches defaultTarget then instead of "universal" task from tasks.json it is built with "template" task.

Maybe it would be better to let CMakeTaskProvider to set by default targets to ['all'] and when executing always query Cmakedriver/cmakeprojcect for requested/default targets?